File: src/logic/emotions.ts
import { Emoticon, EXPENSIVE_RENT_THRESHOLD, type Emoticon as EmoticonType } from '../types/emotion'
import { LogEventKey, type GameState } from '../types/game'

export type BotEmotion = {
  playerId: number
  emoticon: EmoticonType
}

function isBotControlled(state: GameState, playerId: number): boolean {
  const p = state.players[playerId]
  return !!p && (p.isBot || p.botControlled)
}

function playerIdByName(state: GameState, name: unknown): number | null {
  if (typeof name !== 'string') return null
  const idx = state.players.findIndex((p) => p.name === name)
  return idx === -1 ? null : idx
}

export function detectBotEmotions(prev: GameState, next: GameState): BotEmotion[] {
  const newEntries = next.eventLog.slice(prev.eventLog.length)
  const emotions: BotEmotion[] = []

  for (const entry of newEntries) {
    const params = entry.params ?? {}
    switch (entry.key) {
      case LogEventKey.Bankruptcy: {
        const id = playerIdByName(next, params.name)
        if (id !== null && isBotControlled(next, id)) {
          emotions.push({ playerId: id, emoticon: Emoticon.Sad })
        }
        break
      }
      case LogEventKey.PaidRent: {
        const id = playerIdByName(next, params.name)
        if (
          id !== null &&
          isBotControlled(next, id) &&
          typeof params.amount === 'number' &&
          params.amount >= EXPENSIVE_RENT_THRESHOLD
        ) {
          emotions.push({ playerId: id, emoticon: Emoticon.Angry })
        }
        break
      }
      case LogEventKey.MonopolyRent: {
        const id = playerIdByName(next, params.owner)
        if (id !== null && isBotControlled(next, id)) {
          emotions.push({ playerId: id, emoticon: Emoticon.Proud })
        }
        break
      }
      case LogEventKey.TradeAccepted: {
        for (const name of [params.from, params.to]) {
          const id = playerIdByName(next, name)
          if (id !== null && isBotControlled(next, id)) {
            emotions.push({ playerId: id, emoticon: Emoticon.Happy })
          }
        }
        break
      }
      case LogEventKey.DoublesAgain: {
        const id = playerIdByName(next, params.name)
        if (id !== null && isBotControlled(next, id)) {
          emotions.push({ playerId: id, emoticon: Emoticon.Happy })
        }
        break
      }
      default:
        break
    }
  }
  return emotions
}

File: backend/internal/usecase/botusecase/emotions.go
package botusecase

import "monopoly-game-backend/internal/entity"

type BotEmotion struct {
	PlayerID int             `json:"playerId"`
	Emoticon entity.Emoticon `json:"emoticon"`
}

func DetectBotEmotions(prev, next entity.GameState) []BotEmotion {
	start := len(prev.EventLog)
	emotions := make([]BotEmotion, 0)
	if start > len(next.EventLog) {
		return emotions
	}
	for _, entry := range next.EventLog[start:] {
		params := entry.Params
		switch entry.Key {
		case entity.LogEventKeyBankruptcy:
			if id := playerIDByName(next, stringParam(params, "name")); id >= 0 && isBotControlled(next, id) {
				emotions = append(emotions, BotEmotion{PlayerID: id, Emoticon: entity.EmoticonSad})
			}
		case entity.LogEventKeyPaidRent:
			id := playerIDByName(next, stringParam(params, "name"))
			if id >= 0 && isBotControlled(next, id) && numberParam(params, "amount") >= entity.EXPENSIVE_RENT_THRESHOLD {
				emotions = append(emotions, BotEmotion{PlayerID: id, Emoticon: entity.EmoticonAngry})
			}
		case entity.LogEventKeyMonopolyRent:
			if id := playerIDByName(next, stringParam(params, "owner")); id >= 0 && isBotControlled(next, id) {
				emotions = append(emotions, BotEmotion{PlayerID: id, Emoticon: entity.EmoticonProud})
			}
		case entity.LogEventKeyTradeAccepted:
			for _, name := range []string{stringParam(params, "from"), stringParam(params, "to")} {
				if id := playerIDByName(next, name); id >= 0 && isBotControlled(next, id) {
					emotions = append(emotions, BotEmotion{PlayerID: id, Emoticon: entity.EmoticonHappy})
				}
			}
		case entity.LogEventKeyDoublesAgain:
			if id := playerIDByName(next, stringParam(params, "name")); id >= 0 && isBotControlled(next, id) {
				emotions = append(emotions, BotEmotion{PlayerID: id, Emoticon: entity.EmoticonHappy})
			}
		}
	}
	return emotions
}

func isBotControlled(state entity.GameState, playerID int) bool {
	if playerID < 0 || playerID >= len(state.Players) {
		return false
	}
	player := state.Players[playerID]
	return player.IsBot || player.BotControlled
}

func playerIDByName(state entity.GameState, name string) int {
	if name == "" {
		return -1
	}
	for id, player := range state.Players {
		if player.Name == name {
			return id
		}
	}
	return -1
}

func stringParam(params map[string]interface{}, key string) string {
	if params == nil {
		return ""
	}
	value, ok := params[key].(string)
	if !ok {
		return ""
	}
	return value
}

func numberParam(params map[string]interface{}, key string) int {
	if params == nil {
		return 0
	}
	switch value := params[key].(type) {
	case int:
		return value
	case int8:
		return int(value)
	case int16:
		return int(value)
	case int32:
		return int(value)
	case int64:
		return int(value)
	case uint:
		return int(value)
	case uint8:
		return int(value)
	case uint16:
		return int(value)
	case uint32:
		return int(value)
	case uint64:
		return int(value)
	case float32:
		return int(value)
	case float64:
		return int(value)
	default:
		return 0
	}
}
